File: src/longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters

from collections import deque

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        seen = set()
        substr = deque()
        peak = 0
        for char in s:
            while char in seen:
                out = substr.popleft()
                seen.remove(out)
            seen.add(char)
            substr.append(char)
            peak = max(peak, len(substr))
        return peak

if __name__ == "__main__":
    sol = Solution()
    tests = [
        ("abcabcbb", 3),
        ("pwwkew", 3),
        ("bbbbb", 1),
    ]
    for (s, solution) in tests:
        result = sol.lengthOfLongestSubstring(s)
        assert result == solution, \
            f"result {result} != solution {solution}"
    print("✅ All tests passed")

File: n150/longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.rs
// https://leetcode.com/problems/longest-substring-without-repeating-characters
struct Solution;

use std::collections::{VecDeque, HashSet};

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut set = HashSet::with_capacity(s.len());
        let mut window = VecDeque::with_capacity(s.len());
        let mut max = 0;

        for ch in s.chars() {
            while set.contains(&ch) {
                let out = window.pop_front().expect("window and set size mismatch");
                set.remove(&out);
            }
            set.insert(ch);
            window.push_back(ch);
            max = i32::max(max, window.len() as i32);
        }

        max
    }
}

fn main() {
    let tests = vec![
        (
            "abcabcbb".to_string(),
            3,
        ),
        (
            "bbbbbbbb".to_string(),
            1,
        ),
        (
            "pwwkew".to_string(),
            3,
        ),
    ];

    for (s, solution) in tests {
        let result = Solution::length_of_longest_substring(s);
        assert_eq!(result, solution);
    }
}
